### PR TITLE
Provisional Rails 4 updates; specs pass

### PR DESCRIPTION
**Pull request by donaldguy**
- Mongoid 4 uses upstream bson lib enforcing serialization with $oid;
  we unwrap by monkey-patching like promiscous ( http://git.io/cDladA )
- Rspec 3 deprecations quieted

### CODE REVIEW

**Comment by donaldguy** [comment]
(if the branch name doesn't make it clear, mostly this was @michaeldauria in reality)

**Comment by michaeldauria** [comment]
you could have done dg-md-rails-4, you deserve credit too

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -13,4 +13,7 @@
   config.color = true
   config.include DefineConstantHelpers
   config.mock_with :mocha
+  config.expect_with :rspec do |c|
+    c.syntax = [:expect, :should]
+  end
```
Comment (quentindecock) [inline spec/spec_helper.rb:16]: So this is what silences the warnings?

Comment (donaldguy) [inline spec/spec_helper.rb:16]: Yep. The alternative is refactoring all the shoulds to expected_to's. The latter is better, the former is easier :-)

c.f. https://www.relishapp.com/rspec/rspec-expectations/docs/syntax-configuration#explicitly-enable-both-syntaxes

File: lib/elastic_model/bson_monkeypatch.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+else
+  module BSON
+    class ObjectId
+      def as_json(options = {})
+        to_s
+      end
+    end
+  end
+end
+
```
Comment (quentindecock) [inline lib/elastic_model/bson_monkeypatch.rb:18]: remove whitespace

File: spec/spec_helper.rb (hunk 1 above)
Comment (quentindecock) [inline spec/spec_helper.rb:16]: I am all for saving energy here.

File: elastic_model.gemspec [hunk 3]
```diff
@@ -37,5 +37,5 @@ Gem::Specification.new do |spec|
   spec.add_runtime_dependency "mongoid"
   spec.add_runtime_dependency "bson_ext"
   spec.add_runtime_dependency "elasticsearch", "~> 1.0.2"
-  spec.add_runtime_dependency "rails", "~> 3.2"
+  spec.add_runtime_dependency "rails", "~> 4.1.0"
```
Comment (kareemk) [inline elastic_model.gemspec]: so its not backward compatible?

File: lib/elastic_model/bson_monkeypatch.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+if defined? Moped::BSON
+  class Moped::BSON::ObjectId
+    # No {"$oid": "123"}, it's horrible.
+    # TODO Document this shit.
+    def to_json(*args)
+      "\"#{to_s}\""
+    end
```
Comment (quentindecock) [inline lib/elastic_model/bson_monkeypatch.rb:7]: Just for me to understand correctly here.

Before to_json was returning a straight ID as a string.
Now mongoid 4 is returning `{'$oid' => 'abc'}`.
So the way to go around it is to proxy `to_json` to `to_s`?

File: spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb [hunk 5]
```diff
@@ -72,6 +72,8 @@ def es_parent_id
           include ElasticModel::Callbacks
           belongs_to :parent_association
 
+          field :count, :type => Integer
+
```
Comment (quentindecock) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:76]: Why do we need this? We are not updating the test down below based on this change.

File: lib/elastic_model/bson_monkeypatch.rb (hunk 4 above)
Comment (michaeldauria) [inline lib/elastic_model/bson_monkeypatch.rb:7]: Yes, Mongoid 3 used its own BSON library and now Mongoid 4is back to using the BSON library from 10gen: https://github.com/mongodb/bson-ruby/blob/master/lib/bson/object_id.rb#L74-L76

File: spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb (hunk 5 above)
Comment (donaldguy) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:76]: It is used below though. Apparently Mongoid 3 is more forgiving about undefined fields?

(squashes an error under rails 4)

Comment (michaeldauria) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:76]: b/c the test sets `count` and the field is missing, which was breaking in Mongoid 4

Comment (quentindecock) [inline spec/lib/elastic_model/integration/callbacks_with_parent_indices_spec.rb:76]: Oh, so I forgot that guy. Ok.

**Comment by quentindecock** [comment]
LGTM

**Comment by donaldguy** [comment]
What ever became of this?

**Comment by donaldguy** [comment]
I'm guessing this is never happening :smiling_imp:

**Comment by quentindecock** [comment]
An then Donald invented a new discipline: _The Pull Request Archeology_

**Pull request closed without merging**